Move solution examine subscription from DrinkComponent to ExaminableSolutionComponent

iaada opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
Currently solutions get examine text from two components, `DrinkComponent` and `ExaminableSolutionComponent`. ExaminableSolution adds a description of the reagents and the recognizable reagents. Drink adds the text for the current fill level, and calls `HasComp<ExaminableSolutionComponent>` to mode switch between seeing the exact amount or a vague amount.

This PR moves that examine text out of `DrinkComponent` and adds it as a bool to `ExaminableSolutionComponent`. Makes the blocking of examining the solution while closed opt-in. Also moves LocIds to the component.

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
Examining a solution should be tied to `ExaminableSolutionComponent`. This change makes it so that now you can always approximate the fill level of a solution if you can recognize space cola in the solution. It also gives more flexibility by moving an implicit connection between the two components to a bool.

## Technical details
<!-- Summary of code changes for easier review. -->
The structure of the code from `DrinkComponent` is mostly the same moved over to `ExaminableSolutionComponent`. Previously we used a switch with the solution percentage to get different LocIds. Now we use a switch to pass an enum into one LocId. In that LocId we use another switch for each option. I also use a parameterized term to make it more reusable for #

With the physical description of the reagent, we pass the reagent count directly to ftl to lower from three LocId to one. Moved a hardcoded space in recognizable reagents string to ftl.

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

https://github.com/user-attachments/assets/ee0edba1-69b6-40e8-ad15-3324e6234f90

- Opaque drinks like soda cans don't always show you the reagent count.
- Opaque drinks like buckets can have an approximate fill level.
- Non-drinks like fuel tanks can have a fill level.
- Max volume now shows in the examine text.

<img width="448" height="238" alt="image" src="https://github.com/user-attachments/assets/4ea50adb-1f1e-4ff5-91f5-8676993146a5" />

- Max volume doesn't have to show in examine text by using a different LocId.

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
<!--
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->

:cl: aada
- add: You can examine fuel tanks to see how much they're filled.

iaada commented:
I haven't seen `TryStopNukeOpsFromConstantlyFailing` in a while.

Princess-Cheeseballs commented:
> I haven't seen `TryStopNukeOpsFromConstantlyFailing` in a while.

I thought it got fixed, I fear it returning.

iaada commented:
Yeah I didn't mean to steal your PR or anything like that. I just happened to make the same PR, and when you linked yours I saw a few things I could do better in this one.

Princess-Cheeseballs commented on Resources/Locale/en-US/chemistry/solution/components/shared-solution-container-component.ftl:
```diff
@@ -1,9 +1,28 @@
+   *[other] { -solution-vague-fill-level(fillLevel: $fillLevel) }.
+}
+
+-solution-vague-fill-level =
+    { $fillLevel ->
+        [full] [color=white]Full[/color]
+        [mostlyfull] [color=#DFDFDF]Mostly Full[/color]
+        [halffull] [color=#C8C8C8]Half Full[/color]
+        [halfempty] [color=#C8C8C8]Half Empty[/color]
+        [mostlyempty] [color=#A4A4A4]Mostly Empty[/color]
+       *[empty] [color=gray]Empty[/color]
+    }
```
If you're up for it, I would have a second version of this for no-max to use for puddles and such, where the maximum is 1000 currently, but practically it's around 20, currently if you set them to be vague, it always shows "mostly empty" on the puddle. 

Either that or have it so vague with no maximum doesn't show anything in regards to fill level.

iaada commented:
I added examinable solution to `DrinkBase` and the others that you had in your PR. I originally left if off because I had a more complete refactor of drinks with #39221, but that can come later.

I also added a specific loc for puddles. The issue with them is that their max volume is so large that you rarely get different text. At 0-10 you get `Empty`, then from 10-333 you get `MostlyEmpty`. 333+ volume in a single tile is very rare, and doesn't even happen if you break a high-capacity fuel tank in a small room.
<img width="325" height="146" alt="image" src="https://github.com/user-attachments/assets/170df64e-db61-4efa-8b7b-1642021af556" /> <img width="328" height="167" alt="image" src="https://github.com/user-attachments/assets/c5b8e0be-501a-4d00-a492-b7b9da3edfa8" />

iaada commented on Resources/Prototypes/Entities/Objects/Consumable/Food/Containers/condiments.yml:
```diff
@@ -69,6 +71,8 @@
     tags:
       - Packet
+  - type: ExaminableSolution
+    exactVolume: true
```
I guess it's just a personal feeling. I picture the little packet saying "contains 21ml of ketchup" or something like that.

Princess-Cheeseballs replied:
Alright good enough for me. I doubt people using bbq packets to exactly measure reagents will ever matter. And if in some dark timeline it somehow does matter, it's not hard to change.

Princess-Cheeseballs reviewed (approved):
Code looks good + Good cleanup + Way better implementation by using Datafield defined and Enum based descriptions.

beck-thompson commented on Content.Shared/Chemistry/Components/SolutionManager/ExaminableSolutionComponent.cs:
```diff
@@ -1,14 +1,75 @@
+    ///     If true, the solution can't be examined when this entity is closed.
+    /// </summary>
+    [DataField]
+    public bool Opaque;
```
I'd name this to like `CanExamineClosed`

beck-thompson commented on Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerSystem.Capabilities.cs:
```diff
@@ -175,5 +173,14 @@ public static string ToPrettyString(Solution solution)
     }
 
+    /// <returns>A value between 0 and 100 inclusive.</returns>
+    public static float PercentFull(Solution sol)
```
I think this already exists in the file, either use an override or combine them

<img width="881" height="245" alt="Image" src="https://github.com/user-attachments/assets/2c9fb3b9-659e-470f-97ba-fcc211ad7c36" />

iaada replied:
That method was only used for the switch I took out of `DrinkComponent`. I didn't feel right removing it, but I didn't like using it to get the solution again when I already had the solution I could pass to a static method. Would it be best to just remove it and mention it in the breaking changes?

beck-thompson commented on Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerSystem.cs:
```diff
@@ -845,41 +851,58 @@ private void OnExamineSolution(Entity<ExaminableSolutionComponent> entity, ref E
+        if (examiner == null)
+            return ExaminedVolumeDisplay.HalfFull;
+
+        if (TryComp<MetaDataComponent>(examiner, out var meta)
```
Use `MetaData`

beck-thompson commented on Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerSystem.cs:
```diff
@@ -845,41 +851,58 @@ private void OnExamineSolution(Entity<ExaminableSolutionComponent> entity, ref E
+    }
+
+    // Some spessmen see half full, some see half empty, but always the same one.
+    private ExaminedVolumeDisplay HalfEmptyOrHalfFull(EntityUid? examiner = null)
```
I'm glad you kept this

beck-thompson commented on Resources/Prototypes/Entities/Objects/Specific/Medical/hypospray.yml:
```diff
@@ -22,6 +22,8 @@
     solution: hypospray
   - type: ExaminableSolution
     solution: hypospray
+    heldOnly: true
```
Why is it only on held? Don't hyposprays have an sprite indicator on how full they are?

iaada replied:
Sometimes you make a change to yml at 4am and can't really justify why. I think I saw that the hypopen had it and stopped thinking.

beck-thompson replied:
> Sometimes you make a change to yml at 4am and can't really justify why. I think I saw that the hypopen had it and stopped thinking.

LOL no need to justify or explain it at all I was just wondering your reasoning :laughing:
I don't care either way its up to you. I just wanted to make sure it was a deliberate choice!

beck-thompson commented on Content.Shared/Chemistry/Components/SolutionManager/ExaminableSolutionComponent.cs:
```diff
@@ -1,14 +1,76 @@
+    ///     Examine text for when the solution is empty.
+    /// </summary>
+    [DataField]
+    public LocId LocEmptyVolume = "shared-solution-container-component-on-examine-empty-container";
```
I think you should remove this and just use `ExaminedVolumeDisplay` - I'm a little confused though because I can't seem to get the empty loc string to appear 🤔

<img width="521" height="406" alt="Image" src="https://github.com/user-attachments/assets/bb3696f4-334d-485a-9c63-fcbeab6c8280" />

iaada replied:
`LocEmptyVolume` is used when exiting early if no reagent was found in the solution, so you're right that `enum.Empty` is never reached. This was partially done because that's how it was done before, but `Empty` was kept as an option in the enum to use in the follow-up PR #39363.

A different way to go about it could be to get rid of `LocEmptyVolume`, and when exiting early call `LocVolume, ("fillLevel", ExaminedVolumeDisplay.Empty)`.

beck-thompson replied:
I think that makes sense :thinking:

beck-thompson commented on Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerSystem.cs:
```diff
@@ -791,52 +795,54 @@ private void OnContainerManagerInit(Entity<SolutionContainerManagerComponent> en
 
-        if (string.IsNullOrEmpty(primaryReagent?.Prototype))
+        // If there's no primary reagent, assume the solution is empty
+        if (string.IsNullOrEmpty(primaryReagent?.Prototype) ||
```
Put this inside the `using (args.PushGroup(nameof(ExaminableSolutionComponent)))` so it doesn't change locations when drinking! Besides this looks good.

I'd usually just push this change but I am too tired and I worry I'll mess it up 😆

beck-thompson commented:
I think that makes sense

beck-thompson reviewed (approved):
Looks good to me 🥳 🥳 🥳 !
